src/web3-adapter/alchemySendWebSocket.ts

```diff
@@ -0,0 +1,91 @@
+import SturdyWebSocket from "sturdy-websocket";
+import {
+  isResponse,
+  JsonRpcId,
+  SingleOrBatchRequest,
+  SingleOrBatchResponse,
+  WebSocketMessage,
+} from "../types";
+import { AlchemySendFunction, AlchemySendResult } from "./alchemySend";
+
+export function makeWebSocketSender(ws: SturdyWebSocket): AlchemySendFunction {
+  let resolveFunctionsById: Map<
+    JsonRpcId,
+    (response: AlchemySendResult) => void
+  > = new Map();
+  ws.addEventListener("message", message => {
+    const response: WebSocketMessage = JSON.parse(message.data);
+    if (!isResponse(response)) {
+      return;
+    }
+    const id = getIdFromResponse(response);
+    if (id === undefined) {
+      return;
+    }
+    const resolve = resolveFunctionsById.get(id);
+    if (resolve) {
+      resolveFunctionsById.delete(id);
+      if (
+        !Array.isArray(response) &&
+        response.error &&
+        response.error.code === 429
+      ) {
+        resolve({ type: "rateLimit" });
+      } else {
+        resolve({ response, type: "jsonrpc" });
+      }
+    }
+  });
+  ws.addEventListener("down", () => {
+    const oldResolveFunctionsById = resolveFunctionsById;
+    resolveFunctionsById = new Map();
+    for (const [id, resolve] of oldResolveFunctionsById.entries()) {
+      resolve({
+        type: "networkError",
+        status: 0,
+        message: `WebSocket closed before receiving a response for request with id: ${id}`,
+      });
+    }
+  });
+
+  return request =>
+    new Promise(resolve => {
+      const id = getIdFromRequest(request);
+      if (id !== undefined) {
+        const existingResolve = resolveFunctionsById.get(id);
+        if (existingResolve) {
+          const message = `Another WebSocket request was made with the same id (${id}) before a response was received.`;
```

Comment:
You're right, we can have a separate pool for the two. I think in practice this doesn't matter here because the library code will be making unique ids anyways and the user doesn't get to choose their own ids.